File: LegalDocParser/LegalDocParser/Services/SearchService.cs
using System.Text.RegularExpressions;
using LegalDocParser.Models;
using LegalDocParser.Services.Interfaces;

namespace LegalDocParser.Services;

/// <summary>
/// 변환된 문서 내 키워드 검색 서비스
/// </summary>
public class SearchService : ISearchService
{
    private static readonly TimeSpan RegexTimeout = TimeSpan.FromSeconds(5);

    public Task<IReadOnlyList<SearchResult>> SearchAsync(
        string content,
        string pattern,
        bool useRegex = false,
        bool caseSensitive = false)
    {
        return Task.Run(() => Search(content, pattern, useRegex, caseSensitive));
    }

    private static IReadOnlyList<SearchResult> Search(
        string content,
        string pattern,
        bool useRegex,
        bool caseSensitive)
    {
        if (string.IsNullOrEmpty(content) || string.IsNullOrEmpty(pattern))
            return Array.Empty<SearchResult>();

        var results = new List<SearchResult>();
        var lines = content.Split('\n');

        var regexOptions = RegexOptions.Compiled;
        if (!caseSensitive)
            regexOptions |= RegexOptions.IgnoreCase;

        Regex regex;
        try
        {
            var regexPattern = useRegex ? pattern : Regex.Escape(pattern);
            regex = new Regex(regexPattern, regexOptions, RegexTimeout);
        }
        catch (ArgumentException)
        {
            return Array.Empty<SearchResult>();
        }

        for (int i = 0; i < lines.Length; i++)
        {
            var line = lines[i];
            MatchCollection matches;

            try
            {
                matches = regex.Matches(line);
            }
            catch (RegexMatchTimeoutException)
            {
                continue;
            }

            foreach (Match match in matches)
            {
                results.Add(new SearchResult(
                    MatchedText: match.Value,
                    LineNumber: i + 1,
                    ColumnStart: match.Index,
                    MatchLength: match.Length,
                    ContextLine: line.Trim()
                ));
            }
        }

        return results;
    }
}

File: LegalDocParser/services/search_service.py
import re
import signal
import threading

from models.search_result import SearchResult

_REGEX_TIMEOUT_SECONDS = 5


class _RegexTimeoutError(Exception):
    pass


class SearchService:
    def search(
        self,
        content: str,
        pattern: str,
        use_regex: bool = False,
        case_sensitive: bool = False,
    ) -> list[SearchResult]:
        if not content or not pattern:
            return []

        try:
            flags = 0 if case_sensitive else re.IGNORECASE
            if use_regex:
                compiled = re.compile(pattern, flags)
            else:
                compiled = re.compile(re.escape(pattern), flags)
        except re.error:
            return []

        results: list[SearchResult] = []
        lines = content.splitlines()

        try:
            results = self._search_with_timeout(compiled, lines)
        except _RegexTimeoutError:
            return results

        return results

    def _search_with_timeout(
        self, compiled: re.Pattern, lines: list[str]
    ) -> list[SearchResult]:
        results: list[SearchResult] = []
        container: dict = {"done": False, "error": None}

        def _do_search():
            try:
                for line_number, line in enumerate(lines, start=1):
                    for match in compiled.finditer(line):
                        results.append(SearchResult(
                            matched_text=match.group(),
                            line_number=line_number,
                            column_start=match.start(),
                            match_length=len(match.group()),
                            context_line=line,
                        ))
                container["done"] = True
            except Exception as e:
                container["error"] = e

        thread = threading.Thread(target=_do_search, daemon=True)
        thread.start()
        thread.join(timeout=_REGEX_TIMEOUT_SECONDS)

        if not container["done"] and container["error"] is None:
            raise _RegexTimeoutError(
                f"정규식 검색이 {_REGEX_TIMEOUT_SECONDS}초 제한을 초과했습니다."
            )
        if container["error"] is not None:
            raise _RegexTimeoutError(str(container["error"]))

        return results
